examples/hosts/app-integration/corruption-recovery/src/recoveryAgent.ts

```diff
@@ -0,0 +1,291 @@
+        );
+    }
+
+    public get getRecoveryInfo(): RecoveryInfo {
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+        if (!this._orgContainerId) {
+            throw new Error("Cannot retreive status: Missing container ID.");
+        }
+
+        return {
+            originalContainerId: this._orgContainerId,
+            isContainerCorrupted: this._recoveryInfo.get<boolean>(corruptedKey),
+            agentId: this.agentId,
+            recoveryStatus: this._recoveryStatus,
+            isContainerRecovered: this.recovered,
+            recoveryLog: this._recoveryLog,
+            recoveredContainerId: this._recoveryInfo.get<string>(
+                recoveredContainerKey,
+            ),
+            recoveredBy: this._recoveryInfo.get<string>(recoveredByKey),
+        };
+    }
+
+    public get recoveredContainerId(): string | undefined {
+        return this._recoveryInfo.get<string>(recoveredContainerKey);
+    }
+
+    /* Simulated corruption */
+
+    public markCorrupted(): void {
+        this._recoveryInfo.set(corruptedKey, true);
+    }
+
+    /* Recovery */
+
+    public async recoverDoc(cb: ContainerSetupCb): Promise<string | undefined> {
+        this.setRecoveryStatus("KickedOff");
+        this.log("Kicked off recovery.");
+
+        if (!this.markedForRecovery) {
+            this.log("Marked for forecovery.");
+            await this.markForRecovery();
+        }
+
+        if (this.recovered) {
+            this.setRecoveryStatus("Skipped");
+            this.log("Doc already recovered!");
+            return this.recoveredContainerId;
+        }
+
+        this.log("Volonteering for recovery");
+        await this.volunteerForRecovery();
+        if (this.recovered) {
+            this.releaseRecoveryTask();
+            this.setRecoveryStatus("Skipped");
+            this.log("We volonteered, but doc already recovered!");
+            return this.recoveredContainerId;
+        }
+
+        const fluidContainer = await this.recoverContainer();
+        if (!fluidContainer) {
+            this.releaseRecoveryTask();
+            this.setRecoveryStatus("Error");
+            this.log("Could not recover container.");
+            throw new Error("Lost task during write");
+        }
+
+        cb(fluidContainer);
+
+        const id = await fluidContainer.attach();
+        this._recoveryInfo.set(recoveredContainerKey, id);
+
+        if (!this.haveRecoveryTask()) {
+            this.setRecoveryStatus("Error");
+            this.log("Lock error during recovery.");
+            throw new Error("Lost task during write");
+        }
+
+        await this.setRecovered();
+        this.setRecoveryStatus("Success");
+        this.log("This client recovered container.");
+        this.releaseRecoveryTask();
+        return this.recoveredContainerId;
+    }
+
+    private async volunteerForRecovery(): Promise<void> {
+        return this._taskManager.lockTask(recoveryTaskName);
```

Comment:
This will throw if client lost connection. Is this something code expects? I do not think so.
What happens when such exception happens?

Reply:
I made some wrong assumptions here around usage of task locks. I need to rework this item. The possible outcome could be to land to corrupted state, and request the user to kick off recovery again. I'm guessing that error we get back should be meaningful enough to facilitate that experience.